ports/atmel-samd/supervisor/memory.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "supervisor/memory.h"
+
+#include <stddef.h>
+
+#define CIRCUITPY_SUPERVISOR_ALLOC_COUNT 8
+
+static supervisor_allocation allocations[CIRCUITPY_SUPERVISOR_ALLOC_COUNT];
+// We use uint32_t* to ensure word (4 byte) alignment.
+uint32_t* low_address;
+uint32_t* high_address;
+extern uint32_t _ebss;
+extern uint32_t _estack;
+
+void memory_init(void) {
+    low_address = &_ebss;
+    high_address = &_estack;
+}
+
+void free_memory(supervisor_allocation* allocation) {
+    uint8_t index = 0;
+    bool found = false;
+    for (index = 0; index < CIRCUITPY_SUPERVISOR_ALLOC_COUNT; index++) {
+        found = allocation == &allocations[index];
+        if (found) {
+            break;
+        }
```

Comment:
Yes but I couldn't decide how to convey it since the Python VM isn't running. Maybe an `assert`?

Reply:
Maybe we should have a general "terrible thing happened -- blink the status LED?" loop? Or do we need the LED set up to do that?

Reply:
We could use your "blink variable number of times" code and have some ifdef's for the values. This could be a todo. Right now you could just `while { }` like we do for some other errors, and mark it as a TODO.

Reply:
TODO added. Sounds like we'd want to kick into safe mode.